relay-aws-extension/src/aws_extension.rs

```diff
@@ -0,0 +1,193 @@
+            .post(&url)
+            .header(EXTENSION_NAME_HEADER, EXTENSION_NAME)
+            .json(&map)
+            .send()
+            .map_err(|_| AwsExtensionError)?;
+
+        if res.status() != StatusCode::OK {
+            return Err(AwsExtensionError);
+        }
+
+        // let register_response: RegisterResponse = res.json().unwrap();
+
+        let extension_id = res
+            .headers()
+            .get(EXTENSION_ID_HEADER)
+            .and_then(|h| h.to_str().ok())
+            .map(|h| h.to_string())
+            .ok_or(AwsExtensionError)?;
+
+        self.extension_id = Some(extension_id);
+        relay_log::info!("AWS extension successfully registered");
+        Ok(())
+    }
+
+    fn next_event(&self) -> Result<NextEventResponse, AwsExtensionError> {
+        match self.extension_id.as_ref() {
+            Some(extension_id) => {
+                let url = format!("{}/event/next", self.base_url);
+
+                self.reqwest_client
+                    .get(&url)
+                    .header(EXTENSION_ID_HEADER, extension_id)
+                    .send()
+                    .and_then(|r| r.json())
+                    .map_err(|_| AwsExtensionError)
+            }
+            None => Err(AwsExtensionError),
+        }
+    }
+}
+
+impl Actor for AwsExtension {
+    type Context = Context<Self>;
+
+    fn started(&mut self, context: &mut Self::Context) {
+        relay_log::info!("AWS extension started");
+
+        match self.register() {
+            Ok(_) => context.notify(NextEvent),
+            Err(_) => relay_log::info!("AWS extension registration failed"),
+        }
+    }
+
+    fn stopped(&mut self, _context: &mut Self::Context) {
+        relay_log::info!("AWS extension stopped");
+    }
+}
+
+impl Default for AwsExtension {
+    fn default() -> Self {
+        unimplemented!("register with the SystemRegistry instead")
+    }
+}
+
+impl Supervised for AwsExtension {}
+
+impl SystemService for AwsExtension {}
+
+struct NextEvent;
+
+impl Message for NextEvent {
+    type Result = ();
+}
+
+impl Handler<NextEvent> for AwsExtension {
+    type Result = ();
+
+    fn handle(&mut self, _message: NextEvent, context: &mut Self::Context) -> Self::Result {
+        match self.next_event() {
+            Ok(NextEventResponse::Invoke { request_id, .. }) => {
+                relay_log::debug!("Received INVOKE: request_id {}", request_id);
+                context.notify(NextEvent);
+            }
+            Ok(NextEventResponse::Shutdown {
+                shutdown_reason, ..
+            }) => {
+                relay_log::debug!("Received SHUTDOWN: reason {}", shutdown_reason);
+                // need to kill the whole system here, not sure if this is the right way
+                Controller::from_registry().do_send(Signal(SignalType::Term));
+            }
+            _ => {
+                relay_log::debug!("Next event request failed");
+                context.notify(NextEvent);
```

Comment:
This will keep hammering the API endpoint if it is not available. Is that desired behavior? If so I would ask you to hide this actor behind a config option, today it turns on based on whether some envvar is present.

Reply:
I guess I need to distinguish between 404s and other failures here.